analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs

```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
 
         protected class RuleFactory
         {
-            public Type Type { get; }
             private readonly Func<SymbolicRuleCheck> createInstance;
+            private readonly Func<object> sonarFallbackFactory;
+
+            public Type Type { get; }
 
-            protected RuleFactory(Type type, Func<SymbolicRuleCheck> createInstance)
+            protected RuleFactory(Type type, Func<SymbolicRuleCheck> createInstance, Func<object> sonarFallbackFactory)
             {
                 Type = type;
                 this.createInstance = createInstance;
+                this.sonarFallbackFactory = sonarFallbackFactory;
             }
 
-            public SymbolicRuleCheck CreateInstance(SonarAnalysisContext sonarContext, SyntaxNodeAnalysisContext nodeContext)
+            public SymbolicRuleCheck CreateInstance(IAnalyzerConfiguration configuration, SonarAnalysisContext sonarContext, SyntaxNodeAnalysisContext nodeContext)
             {
+                if (configuration.ForceSonarCfg && sonarFallbackFactory != null)
+                {
+                    return null;
+                }
+
                 var ret = createInstance();
                 ret.Init(sonarContext, nodeContext);
                 return ret;
             }
+
+            public object CreateSonarFallback(IAnalyzerConfiguration configuration) =>
+                configuration.ForceSonarCfg ? sonarFallbackFactory?.Invoke() : null;
+        }
+
+        protected class RuleFactory<TCheck> : RuleFactory
+            where TCheck : SymbolicRuleCheck, new()
+        {
+            public RuleFactory() : this(null) { }
+            public RuleFactory(Func<object> sonarFallbackFactory) : base(typeof(TCheck), () => new TCheck(), sonarFallbackFactory) { }
         }
 
-        private sealed class RuleFactory<TCheck> : RuleFactory
+        protected class RuleFactory<TCheck, TSonarFallback> : RuleFactory<TCheck>
```

Comment:
This chaining makes the constructors actually hard to follow and forces the `RuleFactory<TCheck>` to be aware of the sonar fallback - and it shouldn't know about it.  Let's change the inheritance for both to inherit directly `RuleCheck`.

There's slight duplication on the argument calculation, but it makes it more streight-forward in terms of understandability.
```cs
        protected class RuleFactory<TCheck> : RuleFactory
            where TCheck : SymbolicRuleCheck, new()
        {
            public RuleFactory() : base(typeof(TCheck), () => new TCheck(), null) { }
        }

        protected class RuleFactory<TCheck, TSonarFallback> : RuleFactory
            where TCheck : SymbolicRuleCheck, new()
            where TSonarFallback : new()
        {
            public RuleFactory() : base(typeof(TCheck), () => new TCheck(), () => new TSonarFallback()) { }
        }
```

Reply:
`protected` gives a compiler error (`private protected` too). I changed the class to `private sealed` and made the actors `internal`.

Reply:
I don't see this one in the code - the constructors themselves should be changed and one of them removed.